### PR TITLE
[Refactor] Refactor image_demo and huge_image_demo

### PR DESCRIPTION
**Pull request by jbwang1997**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Refactor image_demo and huge_image_demo

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmrotate/apis/patch/merge_results.py [hunk 1]
```diff
@@ -1,45 +1,62 @@
...
-        raise TypeError('Require the shape of `bboxes` to be (n, 5) or (n, 6),'
-                        f' but get `bboxes` with shape being {bboxes.shape}.')
+        raise TypeError('Require the shape of `bboxes` to be (n, 5), (n, 6)'
+                        'or (n, 8), but get `bboxes` with shape being '
+                        f'{bboxes.shape}.')
     return bboxes
 
 
-def map_masks(masks, offset, new_shape):
+def map_masks(masks: np.ndarray, offset, new_shape):
```
Comment (zytx121) [inline mmrotate/apis/patch/merge_results.py:47]:
```suggestion
def map_masks(masks: np.ndarray, offset: Sequence[int], new_shape: Sequence[int]) -> np.ndarray:
```

File: mmrotate/datasets/transforms/loading.py [hunk 2]
```diff
@@ -2,50 +2,44 @@
...
-        super().__init__(*args, **kwargs)
         self.pad_val = pad_val
-        self.padding_mode = padding_mode
 
     def transform(self, results: dict) -> dict:
         """Transform function to add image meta information.
 
         Args:
             results (dict): Result dict with image array in ``results['img']``
                 and patch positioni n ``results['patch']``.
```
Comment (zytx121) [inline mmrotate/datasets/transforms/loading.py:41]:
```suggestion
                and patch position in ``results['patch']``.
```

**Review by zytx121 — DISMISSED** [review]
LGTM

**Review by liuyanyi — APPROVED** [review]

File: mmrotate/apis/inference.py [hunk 3]
```diff
@@ -1,94 +1,112 @@
...
+
+    if test_pipeline is None:
+        cfg = cfg.copy()
+        test_pipeline = cfg.test_dataloader.dataset.pipeline
+        new_test_pipeline, pipeline_types = [], []
+        for pipeline in test_pipeline:
+            if pipeline['type'] != 'LoadAnnotations' and pipeline[
+                    'type'] != 'LoadPanopticAnnotations':
+                new_test_pipeline.append(pipeline)
+                pipeline_types.append(pipeline['type'])
```
Comment (RangiLyu) [inline mmrotate/apis/inference.py:61]: pipeline_types seems to be unused

File: mmrotate/apis/inference.py [hunk 4]
```diff
@@ -1,94 +1,112 @@
...
-        results,
-        windows[:, :2],
-        img_shape=(width, height),
-        iou_thr=merge_iou_thr,
-        device=device)
-    return results
+
+    if test_pipeline is None:
+        cfg = cfg.copy()
+        test_pipeline = cfg.test_dataloader.dataset.pipeline
```
Comment (RangiLyu) [inline mmrotate/apis/inference.py:55]: What if there is a dataset wrapper?

Comment (jbwang1997) [reply mmrotate/apis/inference.py:55]: I add a new function named `get_test_pipeline_cfg` to handle the special cases of dataset wrappers. mmdetection also has the same problem.

File: demo/huge_image_demo.py [hunk 5]
```diff
@@ -60,19 +57,37 @@ def parse_args():
 
 
 def main(args):
+    # register all modules in mmdet into the registries
```
Comment (zytx121) [inline demo/huge_image_demo.py:60]:
```suggestion
    # register all modules in mmrotate into the registries
```

File: demo/image_demo.py [hunk 6]
```diff
@@ -26,18 +28,34 @@ def parse_args():
 
 
 def main(args):
+    # register all modules in mmdet into the registries
```
Comment (zytx121) [inline demo/image_demo.py:31]:
```suggestion
    # register all modules in mmrotate into the registries
```

**Review by zytx121 — DISMISSED** [review]

File: mmrotate/utils/misc.py [hunk 7]
```diff
@@ -37,3 +39,30 @@ def find_latest_checkpoint(path, suffix='pth'):
...
+
+    dataset_cfg = cfg.test_dataloader.dataset
+    test_pipeline = dataset_cfg.get('pipeline', None)
+    # handle dataset wrapper
+    if test_pipeline is None:
+        if 'dataset' in dataset_cfg:
+            test_pipeline = dataset_cfg.dataset.pipeline
+        # handle dataset wrappers like ConcatDataset
+        elif 'datasets' in dataset_cfg:
+            test_pipeline = dataset_cfg.datasets[0].pipeline
```
Comment (RangiLyu) [inline mmrotate/utils/misc.py:65]: There may be nesting of wrappers. You can call this function recursively to solve this problem.

**Review by RangiLyu — APPROVED** [review]

**Pull request merged**